### PR TITLE
#782 fix change username an rename SessionUserCache

### PR DESCRIPTION
**Pull request by Daniel-Majunke**

### CODE REVIEW

File: src/TrueOrFalse.Frontend.Web/Program.cs [hunk 1]
```diff
@@ -121,8 +129,14 @@
     RequestPath = "/Images"
 });
 
+if (useRedis)
+    Redis.UsingRedisSession(app);
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Program.cs:133]: should be "UseRedisSession"

File: src/TrueOrFalse.Frontend.Web/Redis.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            {
+                options.IdleTimeout = TimeSpan.FromMinutes(Settings.SessionStateTimeoutInMin);
+                options.Cookie.HttpOnly = true;
+                options.Cookie.IsEssential = true;
+            });
+        }
+
+        public static void UsingRedisSession(WebApplication app)
+        {
+            app.UseSession();
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Redis.cs:27]: Remove this function

File: src/TrueOrFalse.Frontend.Web/Program.cs [hunk 3]
```diff
@@ -43,12 +45,18 @@
 
 Settings.Initialize(builder.Configuration);
 
-builder.Services.AddSession(options =>
+var useRedis = builder.Configuration.GetValue<bool>("Redis:Use");
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Program.cs:48]: Use Settings.cs for consistency

File: src/TrueOrFalse.Frontend.Web/Redis.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+namespace TrueOrFalse.View.Web
+{
+    public class Redis
+    {
+        public static void Configuration(IServiceCollection services)
+        {
+            services.AddStackExchangeRedisCache(options =>
+            {
+                options.Configuration =
+                    "localhost:6379";
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Redis.cs:14]: should be configured in appsettings.json

File: src/TrueOrFalse.Frontend.Web/Program.cs [hunk 5]
```diff
@@ -43,12 +45,18 @@
...
+    Redis.Configuration(builder.Services);
+else
 {
-    options.Cookie.IsEssential = true;
-    options.Cookie.HttpOnly = true;
-    options.IdleTimeout = TimeSpan.FromMinutes(Settings.SessionStateTimeoutInMin);
-});
+    builder.Services.AddSession(options =>
+    {
+        options.Cookie.IsEssential = true;
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Program.cs:55]: is not related to session storage, should be same for redis and in memory session handling

File: src/TrueOrFalse.Frontend.Web/Program.cs [hunk 6]
```diff
@@ -43,12 +45,18 @@
...
+        options.Cookie.HttpOnly = true;
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Program.cs:56]: see above

File: src/Docker/Meilisearch/docker-compose.meilisearch.redis.yml
Comment (tj-mai) [inline src/Docker/Meilisearch/docker-compose.meilisearch.redis.yml:1]: should be deleted

File: src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/Components/Topic/TopicDataManager.cs [hunk 7]
```diff
@@ -95,8 +96,8 @@ private SearchTopicItem FillMiniTopicItem(CategoryCacheItem topic)
                 );
             }).ToArray(),
             topic.IsStartPage(),
-            _sessionUser.User != null && _sessionUser.UserId == topic.Creator?.Id,
-            _sessionUser.User != null && _permissionCheck.CanDelete(topic),
+            sessionUser != null && _sessionUser.UserId == topic.Creator?.Id,
+            sessionUser != null && _permissionCheck.CanDelete(topic),
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/Components/Topic/TopicDataManager.cs:100]: null check should be handled in CanDelete(..)
remove "sessionUser != null"

File: src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/Components/Topic/TopicDataManager.cs [hunk 8]
```diff
@@ -55,6 +55,7 @@ private SearchTopicItem FillMiniTopicItem(CategoryCacheItem topic)
         ImageMetaData imageMetaData,
         KnowledgeSummary knowledgeSummary)
     {
+        var sessionUser = _sessionUser.UserId == 0 ? null : _sessionUser.User;
         return new TopicDataResult(
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/Components/Topic/TopicDataManager.cs:59]: It should be visible, what vars are populated. Use something like

return new TopicDataResult(
  varName: asdfasdf, 
  ..

File: src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/Components/Topic/TopicDataManager.cs [hunk 9]
```diff
@@ -95,8 +96,8 @@ private SearchTopicItem FillMiniTopicItem(CategoryCacheItem topic)
                 );
             }).ToArray(),
             topic.IsStartPage(),
-            _sessionUser.User != null && _sessionUser.UserId == topic.Creator?.Id,
-            _sessionUser.User != null && _permissionCheck.CanDelete(topic),
+            sessionUser != null && _sessionUser.UserId == topic.Creator?.Id,
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/Components/Topic/TopicDataManager.cs:99]: IsCreator logic should be encapsulated.

File: src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/Pages/Question/QuestionLandingPageController.cs [hunk 10]
```diff
@@ -104,7 +104,7 @@ public QuestionPageResult GetQuestionPage([FromRoute] int id)
                 Solution = q.Solution,
                 IsCreator = q.Creator.Id == _sessionUser.UserId,
                 IsInWishknowledge = _sessionUser.IsLoggedIn &&
-                                    q.IsInWishknowledge(_sessionUser.UserId, _sessionUserCache),
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/Pages/Question/QuestionLandingPageController.cs:107]: fix formating

File: src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/VueSessionUser.cs [hunk 11]
```diff
@@ -45,7 +45,7 @@ public record struct ActivityPoints(
     public CurrentUserData GetCurrentUserData()
     {
         var type = UserType.Anonymous;
-        var user = _sessionUser.User;
+        var user = _sessionUser.UserId == 0 ? null : _sessionUser.User;
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/Views/ApiVueApp/VueSessionUser.cs:48]: use encapsulated logic like IsLoggedIn()

File: src/TrueOrFalse.Frontend.Web/appsettings.json [hunk 12]
```diff
@@ -32,5 +32,8 @@
   "Email": {
     "EmailFrom": "team@memucho.de",
     "EmailToMemucho": "team@memucho.de"
-  }
+  },
+  "Redis": {
+    "Use": true 
+  }
```
Comment (tj-mai) [inline src/TrueOrFalse.Frontend.Web/appsettings.json:38]: Add Hostname

File: src/TrueOrFalse/Tools/Cache/SessionUserCache/ExtendedUserCache.cs [hunk 13]
```diff
@@ -11,16 +12,35 @@ public class SessionUserCache(
...
+        {
+            Logg.r.Error($"userid < 1, {Environment.StackTrace} ");
+            throw new Exception("userId does not exist");
+        }
+
+        var extendedUser = GetItem(userId);
+        if (extendedUser != null)
+            return extendedUser;
+
+        var user = _userReadingRepo.GetById(userId);
```
Comment (tj-mai) [inline src/TrueOrFalse/Tools/Cache/SessionUserCache/ExtendedUserCache.cs:35]: User should be loaded from Cache

File: src/TrueOrFalse/Tools/Cache/SessionUserCache/ExtendedUserCache.cs [hunk 14]
```diff
@@ -81,23 +101,27 @@ public void Update(User user)
...
+            var sessionUserCacheItem = GetItem(user.Id);
 
-        if (GetItem(user.Id) is not null)
-            Log.Error("Expected cache item to be null. Needs a check!");
+            if (sessionUserCacheItem != null)
+                return sessionUserCacheItem;
 
-        cacheItem.Populate(user);
-        AddToCache(cacheItem);
+            var cacheItem = CreateSessionUserItemFromDatabase(user);
```
Comment (tj-mai) [inline src/TrueOrFalse/Tools/Cache/SessionUserCache/ExtendedUserCache.cs:119]: Not From Database, but from Cache

File: src/TrueOrFalse.Frontend.Web/Program.cs (hunk 1 above)
Comment (Daniel-Majunke) [reply src/TrueOrFalse.Frontend.Web/Program.cs:133]: removed

**Review by tj-mai — APPROVED** [review]

**Pull request merged**